oidc/clientassertion/client_assertion.go

```diff
@@ -0,0 +1,206 @@
+	ErrMissingFuncNow         = errors.New("missing now func; please use New()")
+)
+
+// New sets up a new ClientAssertion to sign private key JWTs
+func New(clientID string, audience []string, opts ...Option) (*ClientAssertion, error) {
+	a := &ClientAssertion{
+		clientID: clientID,
+		audience: audience,
+		headers:  make(map[string]string),
+		genID:    uuid.GenerateUUID,
+		now:      time.Now,
+	}
+	for _, opt := range opts {
+		opt(a)
+	}
+	if err := a.Validate(); err != nil {
+		return nil, fmt.Errorf("new client assertion validation error: %w", err)
+	}
+	return a, nil
+}
+
+// ClientAssertion signs a JWT with either a private key or a secret
+type ClientAssertion struct {
+	// for JWT claims
+	clientID string
+	audience []string
+	headers  map[string]string
+
+	// for signer
+	alg jose.SignatureAlgorithm
+	// key may be any key type that jose.SigningKey accepts for its Key
+	key any
+	// secret may be used instead of key
+	secret string
+
+	// these are overwritten for testing
+	genID func() (string, error)
+	now   func() time.Time
+}
+
+// Validate validates the expected fields
+func (c *ClientAssertion) Validate() error {
+	var errs []error
+	if c.genID == nil {
+		errs = append(errs, ErrMissingFuncIDGenerator)
+	}
+	if c.now == nil {
+		errs = append(errs, ErrMissingFuncNow)
+	}
+	// bail early if any internal func errors
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+
+	if c.clientID == "" {
+		errs = append(errs, ErrMissingClientID)
+	}
+	if len(c.audience) == 0 {
+		errs = append(errs, ErrMissingAudience)
+	}
+	if c.alg == "" {
+		errs = append(errs, ErrMissingAlgorithm)
+	}
+	if c.key == nil && c.secret == "" {
+		errs = append(errs, ErrMissingKeyOrSecret)
+	}
+	if c.key != nil && c.secret != "" {
+		errs = append(errs, ErrBothKeyAndSecret)
+	}
+	return errors.Join(errs...)
+}
+
+// SignedToken returns a signed JWT in the compact serialization format
+func (c *ClientAssertion) SignedToken() (string, error) {
+	if err := c.Validate(); err != nil {
+		return "", err
+	}
+	builder, err := c.builder()
+	if err != nil {
+		return "", err
+	}
+	token, err := builder.CompactSerialize()
+	if err != nil {
+		return "", fmt.Errorf("failed to sign token: %w", err)
+	}
+	return token, nil
+}
+
+func (c *ClientAssertion) builder() (jwt.Builder, error) {
+	signer, err := c.signer()
+	if err != nil {
+		return nil, err
+	}
+	id, err := c.genID()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate token id: %w", err)
+	}
+	claims := c.claims(id)
+	return jwt.Signed(signer).Claims(claims), nil
+}
+
+func (c *ClientAssertion) signer() (jose.Signer, error) {
+	sKey := jose.SigningKey{
+		Algorithm: c.alg,
+	}
+
+	// Validate() ensures these are mutually exclusive
+	if c.secret != "" {
+		sKey.Key = []byte(c.secret)
+	}
+	if c.key != nil {
+		sKey.Key = c.key
+	}
+
+	sOpts := &jose.SignerOptions{
+		ExtraHeaders: make(map[jose.HeaderKey]interface{}, len(c.headers)),
+	}
+	// note: extra headers can override "kid"
```

Comment:
Do we want to prevent this?

Reply:
I don't think it actually hurts anything to let folks do this, but it could potentially cause confusion?

Reply:
I think we probably should prevent it, so that it can only be set with the `WithKeyID` option. Does that make sense?